src/flusher.cpp

```diff
@@ -0,0 +1,906 @@
+ *    notice, this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *  * Neither the name of NVIDIA CORPORATION nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "flusher.hpp"
+
+struct flusher_qp_info * flusher_qp=NULL;
+
+//-------------------------------- STATIC ------------------------------------
+static int * flsign_h;
+static gds_flusher_buf flread_d;
+static gds_flusher_buf flack_d;
+static int flusher_value=0;
+static pthread_t flusher_thread;
+static int gds_use_flusher = -1;
+static int local_gpu_id=0;
+
+static const char * flusher_int_to_str(int flusher_int)
+{
+    if(flusher_int == GDS_FLUSHER_NONE)
+        return "No flusher";
+    else if(flusher_int == GDS_FLUSHER_NATIVE)
+        return "GPU native flusher";
+    else if(flusher_int == GDS_FLUSHER_CPU)
+        return "CPU thread";
+    else if(flusher_int == GDS_FLUSHER_NIC)
+        return "NIC RDMA PUT";
+    else
+        return "Unknown";
+}
+
+static inline bool gds_flusher_service_active() {
+    if(gds_use_flusher == GDS_FLUSHER_CPU || gds_use_flusher == GDS_FLUSHER_NIC)
+        return true;
+    else
+        return false;
+}
+#define CHECK_FLUSHER_SERVICE()                                                 \
+    if(gds_flusher_service_active() == false)                                   \
+    {                                                                           \
+        gds_dbg("Flusher service not active (%d)\n", gds_use_flusher);         \
+        goto out;                                                               \
+    }
+
+#define ROUND_TO(V,PS) ((((V) + (PS) - 1)/(PS)) * (PS))
+
+bool gds_use_native_flusher()
+{
+    if(gds_use_flusher == GDS_FLUSHER_NATIVE)
+        return true;
+
+    return false;
+}
+
+static int gds_flusher_pin_buffer(gds_flusher_buf * fl_mem, size_t req_size, int type_mem)
+{
```

Comment:
why do you need a new memory allocation/registration function? why not using/extending those already here?